File: erpnext/stock/report/stock_issue_report/stock_issue_report.py
# # Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# # For license information, please see license.txt


from __future__ import unicode_literals
import frappe
from frappe.utils import flt, cint,add_days, cstr, flt, getdate, nowdate, rounded, date_diff

def execute(filters=None):
    columns = get_columns(filters)
    data = get_data(filters)
    return columns, data


def get_columns(filters):
    cols = [
        ("Date") + ":date:100",
        ("Material Code") + ":data:110",
        ("Material Name")+":data:120",
        ("Material Group")+":data:120",
        ("Material Sub Group")+":data:150",
        ("UoM") + ":data:50",
        ("Qty")+":data:50",
        ("Valuation Rate")+":data:120",
        ("Amount")+":Currency:110",
    ]
    if filters.purpose == "Material Issue" and not filters.accumulate :
        cols.append(("Cost Center")+":data:170")
    if filters.accumulate == 1:
        cols.remove(("Valuation Rate")+":data:120")

    if filters.purpose == "Material Transfer" and not filters.accumulate:
        cols.append(("Warehouse")+":data:170")
        cols.append(("Stock Entry")+":Link/Stock Entry:170")
    return cols

def get_data(filters):
    cond = ''
    if filters.get("warehouse"):
        cond += " AND sed.s_warehouse = \'" + str(filters.warehouse) + "\'"
    if filters.get("from_date") and filters.get("to_date"):
        cond += " AND se.posting_date BETWEEN \'" + str(filters.from_date) + "\' and \'"+ str(filters.to_date) + "\'"
    data = []
    
    if filters.purpose == 'Material Transfer':
        if  filters.accumulate:
            data =frappe.db.sql("""SELECT se.posting_date, sed.item_code, sed.item_name, 
                    (select item_group from `tabItem` where item_code = sed.item_code ) as item_group ,
                    (select item_sub_group from `tabItem` where item_code = sed.item_code ) as item_sub_group, sed.uom, sum(sed.qty) as qty, 
                    sum(sed.amount) as amount
                FROM `tabStock Entry` se inner join `tabStock Entry Detail` sed
                on se.name = sed.parent 
                where se.docstatus = 1
                and se.purpose = 'Material Transfer'
                and sed.item_code = '{}' {}""".format(filters.item_code, cond))
        else:
            if filters.get("item_code"):
                cond += " AND sed.item_code = \'" + str(filters.item_code) + "\'"
            data = frappe.db.sql("""
                            SELECT se.posting_date, sed.item_code, sed.item_name, 
                                (select item_group from `tabItem` where item_code = sed.item_code ) as item_group ,
                                (select item_sub_group from `tabItem` where item_code = sed.item_code ) as item_sub_group,
                                sed.uom, sed.qty, 
                                sed.valuation_rate, sed.amount
                            FROM `tabStock Entry` se inner join `tabStock Entry Detail` sed
                            on se.name = sed.parent 
                            where se.docstatus = 1 
                            and se.purpose = 'Material Transfer'
                            {} """.format(cond))
                    
    elif filters.purpose == 'Material Issue':
        if  filters.accumulate:
            data =frappe.db.sql("""SELECT se.posting_date, sed.item_code, sed.item_name, 
                    (select item_group from `tabItem` where item_code = sed.item_code ) as item_group ,
                    (select item_sub_group from `tabItem` where item_code = sed.item_code ) as item_sub_group, sed.uom, sum(sed.qty) as qty, 
                    sum(sed.amount) as amount
                FROM `tabStock Entry` se inner join `tabStock Entry Detail` sed
                on se.name = sed.parent 
                where se.docstatus = 1
                and se.purpose = 'Material Issue'
                and sed.item_code = '{}' {}""".format(filters.item_code, cond))
        else:
            if filters.get("item_code"):
                cond += " AND sed.item_code = \'" + str(filters.item_code) + "\'"
            data = frappe.db.sql("""
                            SELECT se.posting_date, sed.item_code, sed.item_name, 
                                (select item_group from `tabItem` where item_code = sed.item_code ) as item_group ,
                                (select item_sub_group from `tabItem` where item_code = sed.item_code ) as item_sub_group,
                                sed.uom, sed.qty, 
                                sed.valuation_rate, sed.amount
                            FROM `tabStock Entry` se inner join `tabStock Entry Detail` sed
                            on se.name = sed.parent 
                            where se.docstatus = 1 
                            and se.purpose = 'Material Issue'
                            {} """.format(cond))
            
            #  if filters.get("item_code"):
            #     cond += " AND sed.item_code = \'" + str(filters.item_code) + "\'"
            # data = """SELECT se.posting_date, sed.item_code, sed.item_name, 
            #     i.item_group, i.item_sub_group, sed.uom, sed.qty, 
            #     sed.valuation_rate,sed.amount, sed.cost_center, se.name 
            # FROM `tabStock Entry` se, `tabStock Entry Detail` sed, `tabItem` i
            # WHERE se.name = sed.parent 
            # AND  se.docstatus = 1 
            # AND se.purpose = 'Material Issue'
            # AND i.name = sed.item_code"""
    
    return data

# from __future__ import unicode_literals
# import frappe
# from frappe.utils import flt, cint,add_days, cstr, flt, getdate, nowdate, rounded, date_diff

# def execute(filters=None):
# 	columns = get_columns(filters)
# 	data = get_data(filters)
# 	return columns, data


# def get_columns(filters):
# 	cols = [
# 		("Date") + ":date:100",
# 		("Material Code") + ":data:110",
# 		("Material Name")+":data:120",
# 		("Material Group")+":data:120",
# 		("Material Sub Group")+":data:150",
# 		("UoM") + ":data:50",
# 		("Qty")+":data:50",
# 		("Valuation Rate")+":data:120",
# 		("Amount")+":Currency:110",
# 	]
# 	if filters.purpose == "Material Issue":
# 		cols.append(("Cost Center")+":data:170")


# 	if filters.purpose == "Material Transfer":
# 		cols.append(("Warehouse")+":data:170")
# 	cols.append(("Stock Entry")+":Link/Stock Entry:170")
# 	return cols

# def get_data(filters):
# 	if filters.purpose == 'Material Transfer':
# 		data = """SELECT se.posting_date, sed.item_code, sed.item_name, 
# 				i.item_group, i.item_sub_group, sed.uom, sed.qty, 
# 				sed.valuation_rate,sed.amount, sed.t_warehouse, se.name 
# 			FROM `tabStock Entry` se, `tabStock Entry Detail` sed, `tabItem` i
# 			WHERE se.name = sed.parent 
# 			AND  se.docstatus = 1 
# 			AND se.purpose = 'Material Transfer'
# 			AND i.name = sed.item_code"""
# 	elif filters.purpose == 'Material Issue':
# 		data = """SELECT se.posting_date, sed.item_code, sed.item_name, 
# 				i.item_group, i.item_sub_group, sed.uom, sed.qty, 
# 				sed.valuation_rate,sed.amount, sed.cost_center, se.name 
# 			FROM `tabStock Entry` se, `tabStock Entry Detail` sed, `tabItem` i
# 			WHERE se.name = sed.parent 
# 			AND  se.docstatus = 1 
# 			AND se.purpose = 'Material Issue'
# 			AND i.name = sed.item_code"""
# 	if filters.get("warehouse"):
# 		data += " AND sed.s_warehouse = \'" + str(filters.warehouse) + "\'"
# 	if filters.get("item_code"):
# 		data += " AND sed.item_code = \'" + str(filters.item_code) + "\'"
# 	if filters.get("from_date") and filters.get("to_date"):
# 		data += " AND se.posting_date BETWEEN \'" + str(filters.from_date) + "\' and \'"+ str(filters.to_date) + "\'"
# 	return frappe.db.sql(data)

# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

File: erpnext/stock/report/stock_issue_report/stock_issue_report.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Stock Issue Report"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": ("Company"),
			"fieldtype": "Link",
 			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname":"purpose",
			"label": __("Purpose"),
			"fieldtype": "Select",
			"width": "80",
			"options": ["Material Issue", "Material Transfer"],
			"reqd": 1
		},
		{
            "fieldname": "cost_center",
            "label": __("Parent Cost Center"),
			"fieldtype": "Link",
            "width": "80",
			"options": "Cost Center",
			"get_query": function() {
                                return {
                                        'doctype': "Cost Center",
                                        'filters': [
                                                ['is_disabled', '!=', '1'],
                                                ['is_group', '=', '1']
                                        ]
                                }
                        },
			"on_change": function(){
				var cost_center = frappe.query_report.filters_by_name.cost_center.get_value();
				// frappe.msgprint(branch)
				frappe.call({
					method:"erpnext.stock.report.stock_balance_report.stock_balance_report.get_warehouse",
					args:{"cost_center":cost_center},
					callback: function(r){
						// console.log(r.message)
						// frappe.query_report.filters_by_name.warehouse.set_option(r.message)					
						if(r.message)
						{
							options = []
							for (i = 0; i < r.message.length; i++) { 
								options[i]= r.message[i].name
							}
							console.log(options)
							frappe.query_reports["Stock Issue Report"].filters[6].options = options
							frappe.query_report.filters_by_name.warehouse.refresh();
							frappe.query_report.refresh();
                                                        // **I have set options dynamically to the below select fieldtype but I need to refresh that field to show that new options.**
							// console.log(frappe.query_reports["Stock Balance Report"].filters[4].options)
						}
					}
				});
			}
		},
		{
			"fieldname": "branch",
			"label": ("Branch"),
			"fieldtype": "Link",
			"options": "Cost Center",
			"get_query": function() {
					var cost_center = frappe.query_report.filters_by_name.cost_center.get_value();
					var company = frappe.query_report.filters_by_name.company.get_value();
					if(cost_center!= 'Natural Resource Development Corporation Ltd - NRDCL')
					{
							return {"doctype": "Cost Center", "filters": {"company": company, "is_disabled": 0, "parent_cost_center": cost_center}}
					}
					else
					{
							return {"doctype": "Cost Center", "filters": {"company": company, "is_disabled": 0, "is_group": 0}}
					}
			},
			"on_change": function(){
				var branch = frappe.query_report.filters_by_name.branch.get_value();
				// frappe.msgprint(branch)
				frappe.call({
					method:"erpnext.stock.report.stock_balance_report.stock_balance_report.get_warehouse",
					args:{"cost_center":branch},
					callback: function(r){
						// console.log(r.message)
						// frappe.query_report.filters_by_name.warehouse.set_option(r.message)					
						if(r.message)
						{
							options = []
							for (i = 0; i < r.message.length; i++) { 
								options[i]= r.message[i].name
							}
							console.log(options)
							frappe.query_reports["Stock Issue Report"].filters[6].options = options
							frappe.query_report.filters_by_name.warehouse.refresh();
							frappe.query_report.refresh();
                                                        // **I have set options dynamically to the below select fieldtype but I need to refresh that field to show that new options.**
							// console.log(frappe.query_reports["Stock Balance Report"].filters[4].options)
						}
					}
					/*	console.log(r.message)
						$.each(r.message, function(i, data){
							$('.input-with-feedback').append(new Option(data.name))
						});
					frappe.query_reports.filters[1].refresh();
					} */
				});
			}
		},
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"width": "80",
			"default": sys_defaults.year_start_date,
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"width": "80",
			"default": frappe.datetime.get_today()
		},
		{
			"fieldname": "warehouse",
			"label": __("Warehouse"),
			"fieldtype": "Select",
			"width": "80",
			"options": [""]
		},

		{
			"fieldname": "item_code",
			"label": __("Material Code"),
			"fieldtype": "Link",
			"width": "80",
			"options": "Item"
		},
		{
			"fieldname": "lot_number",
			"label": __("Lot Number"),
			"fieldtype": "Link",
			"width": "100",
			"options": "Lot List"
		}

	]
}
